mlir/lib/Analysis/Presburger/Coalesce.cpp

```diff
@@ -0,0 +1,909 @@
+    addInequalities(bPrime, constraints_b);
+    shift(t, 1);
+    Simplex simp(bPrime);
+    simp.addEquality(t);
+    if (simp.isEmpty()) {
+      // If bPrime is empty, t is shifted back and reclassified as redundant.
+      shift(t, -1);
+      info_a.redundant.push_back(t);
+      info_a.cut.erase(info_a.cut.begin() + l);
+    } else {
+      // Otherwise, all cut constraints of B are considered. Of those, only the
+      // ones, that actually define bPrime are considered. So the ones, that
+      // actually "touche" the polytope bPrime. Those are wrapped around t(x) +
+      // 1 to include a.
+      for (size_t k = 0; k < info_b.cut.size(); k++) {
+        SmallVector<int64_t, 8> curr1 = arrayRefToSmallVector(info_b.cut[k]);
+        Simplex simp2(bPrime);
+        simp2.addEquality(curr1);
+        // TODO: "touching" the polytope is currently defined by adding the
+        // constraint to be checked as an equality and then looking whether this
+        // polytope is empty or not. This requires, that the point at which the
+        // constraint touches the polytope, is an integer point. It is not
+        // entirely clear yet, whether this is sufficient.
+        if (!simp2.isEmpty()) {
+          auto result = wrapping(a, t, curr1);
+          if (!result) {
+            return false;
+          }
+          wrapped.push_back(result.getValue());
+        }
+      }
+    }
+  }
+
+  FlatAffineConstraints new_set(b.getNumDimIds(), b.getNumSymbolIds());
+  // If all the wrappings were succesfull, the two polytopes can be replaced by
+  // a polytope with all of the redundant constraints and the wrapped
+  // constraints.
+  addInequalities(new_set, info_a.redundant);
+  addInequalities(new_set, info_b.redundant);
+  for (size_t k = 0; k < wrapped.size(); k++) {
+    new_set.addInequality(wrapped[k]);
+  }
+  // Additionally for every remaining cut constraint t of a, t + 1 >= 0 is
+  // added.
+  for (size_t k = 0; k < info_a.cut.size(); k++) {
+    SmallVector<int64_t, 8> curr = arrayRefToSmallVector(info_a.cut[k]);
+    shift(curr, 1);
+    new_set.addInequality(curr);
+  }
+  addCoalescedBasicSet(basicSetVector, i, j, new_set);
+  return true;
+}
+
+bool stickingOut(const SmallVector<ArrayRef<int64_t>, 8> &cut,
+                 const FlatAffineConstraints &bs) {
+  Simplex simp(bs);
+  // for every cut constraint t, compute the optimum in the direction of cut. If
+  // the optimum is < -2, the polytopee doesn't stick too much out of the other
+  // one.
+  for (size_t k = 0; k < cut.size(); k++) {
+    ArrayRef<int64_t> curr = cut[k];
+    auto result = simp.computeOptimum(Simplex::Direction::Down, curr);
+    if (!result) {
+      return false;
+    }
+    auto res = result.getValue();
+    if (res.num <= -2 * res.den) {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool mlir::sameConstraint(ArrayRef<int64_t> c1, ArrayRef<int64_t> c2) {
+  Fraction ratio(0, 1);
+  if (c1.size() != c1.size()) {
+    return false;
```

Comment:
Should this rather be an assertion failure? Can this ever happen normally?

Reply:
This cannot happen within my code, but as I define this for the whole mlir namespace, I decided to cover this case.

I will change this to an assertion, as someone who is actually comparing constraints with different sizes might be doing something else wrong.

Note: Depending on our representation of existential dimensions and int divisions, this case might happen actually AFAIU.